pkg/kubelet/server/metrics/metrics.go

```diff
@@ -64,6 +64,19 @@ var (
 		},
 		[]string{"method", "path", "server_type", "long_running"},
 	)
+	// VolumeStatCalDuration tracks the duration in seconds to calculate volume stats.
+	// this metric is mainly for comparison between fsquota monitoring and `du` for disk usage.
+	VolumeStatCalDuration = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Subsystem: kubeletSubsystem,
+			Name:      "volume_metric_collection_duration_seconds",
+			Help:      "Duration in seconds to calculate volume stats",
+			// Use DefBuckets for now, will customize the buckets if necessary.
+			Buckets:        metrics.DefBuckets,
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{},
+	)
```

Comment:
Should this have any labels? i.e - getting stats can vary between volume backends and do we need to identify which storage provider we are dealing with?

Reply:
If we add labels, there would be too many metrics if we use histogram vec here.

So I prefer to add labels if we change it to a Gauge that volume_metric_collection_duration_seconds for the latest running duration of getMetrics.

Histogram for all volumes or the latest running Gauge for each volume. Or both.

Reply:
Gauge for each volume is like below:
code: https://github.com/kubernetes/kubernetes/compare/master...pacoxu:add-metrics-volume-stats-cal-3?expand=1
```
# TYPE kubelet_volume_metric_collection_duration_seconds gauge
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-apiserver-5b68b6b54-qzt8q",pod_namespace="calico-apiserver",volume_name="calico-apiserver-certs"} 0.000113258
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-apiserver-5b68b6b54-qzt8q",pod_namespace="calico-apiserver",volume_name="kube-api-access-lpt2s"} 6.966e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-apiserver-5b68b6b54-tt2ww",pod_namespace="calico-apiserver",volume_name="calico-apiserver-certs"} 0.000192778
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-apiserver-5b68b6b54-tt2ww",pod_namespace="calico-apiserver",volume_name="kube-api-access-htpq2"} 0.000204466
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-kube-controllers-59c45ff85c-x2gts",pod_namespace="calico-system",volume_name="kube-api-access-znzxf"} 9.0457e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="cni-bin-dir"} 2.168e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="cni-log-dir"} 1.3446e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="cni-net-dir"} 2.0964e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="felix-certs"} 0.000190311
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="flexvol-driver-host"} 2.585e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="kube-api-access-5gpf8"} 0.000121745
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="lib-modules"} 2.5808e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="policysync"} 2.417e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="typha-ca"} 0.000178574
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="var-lib-calico"} 1.9599e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="var-run-calico"} 2.0434e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-node-d2jpf",pod_namespace="calico-system",volume_name="xtables-lock"} 2.0314e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-typha-65cdbb598d-644x5",pod_namespace="calico-system",volume_name="kube-api-access-j4smz"} 0.000233179
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-typha-65cdbb598d-644x5",pod_namespace="calico-system",volume_name="typha-ca"} 0.000312874
kubelet_volume_metric_collection_duration_seconds{pod_name="calico-typha-65cdbb598d-644x5",pod_namespace="calico-system",volume_name="typha-certs"} 0.000138765
kubelet_volume_metric_collection_duration_seconds{pod_name="coredns-8549969ffd-8qwz6",pod_namespace="kube-system",volume_name="config-volume"} 0.000133681
kubelet_volume_metric_collection_duration_seconds{pod_name="coredns-8549969ffd-8qwz6",pod_namespace="kube-system",volume_name="kube-api-access-h45nt"} 7.2941e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="coredns-8549969ffd-nv5wh",pod_namespace="kube-system",volume_name="config-volume"} 5.181e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="coredns-8549969ffd-nv5wh",pod_namespace="kube-system",volume_name="kube-api-access-8xpcp"} 9.6276e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="etcd-daocloud",pod_namespace="kube-system",volume_name="etcd-certs"} 1.873e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="etcd-daocloud",pod_namespace="kube-system",volume_name="etcd-data"} 9.5661e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-apiserver-daocloud",pod_namespace="kube-system",volume_name="ca-certs"} 5.935e-06
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-apiserver-daocloud",pod_namespace="kube-system",volume_name="etc-pki"} 6.812e-06
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-apiserver-daocloud",pod_namespace="kube-system",volume_name="k8s-certs"} 0.000521161
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-controller-manager-daocloud",pod_namespace="kube-system",volume_name="ca-certs"} 6.6133e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-controller-manager-daocloud",pod_namespace="kube-system",volume_name="etc-pki"} 2.4184e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-controller-manager-daocloud",pod_namespace="kube-system",volume_name="flexvolume-dir"} 4.9894e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-controller-manager-daocloud",pod_namespace="kube-system",volume_name="k8s-certs"} 3.6707e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-controller-manager-daocloud",pod_namespace="kube-system",volume_name="kubeconfig"} 2.9778e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-proxy-n4mdq",pod_namespace="kube-system",volume_name="kube-api-access-p57q5"} 0.00011465
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-proxy-n4mdq",pod_namespace="kube-system",volume_name="kube-proxy"} 0.000141253
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-proxy-n4mdq",pod_namespace="kube-system",volume_name="lib-modules"} 9.1851e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-proxy-n4mdq",pod_namespace="kube-system",volume_name="xtables-lock"} 7.927e-06
kubelet_volume_metric_collection_duration_seconds{pod_name="kube-scheduler-daocloud",pod_namespace="kube-system",volume_name="kubeconfig"} 1.6952e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="nfs-client-provisioner-599cffff7d-ktr7f",pod_namespace="default",volume_name="kube-api-access-7njp7"} 9.0973e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="nfs-client-provisioner-599cffff7d-ktr7f",pod_namespace="default",volume_name="nfs-client-root"} 0.001036602
kubelet_volume_metric_collection_duration_seconds{pod_name="pod-test",pod_namespace="default",volume_name="data-test"} 0.000141978
kubelet_volume_metric_collection_duration_seconds{pod_name="pod-test",pod_namespace="default",volume_name="kube-api-access-z6jhc"} 0.000270384
kubelet_volume_metric_collection_duration_seconds{pod_name="pod-test1",pod_namespace="default",volume_name="data-test"} 3.2148e-05
kubelet_volume_metric_collection_duration_seconds{pod_name="pod-test1",pod_namespace="default",volume_name="kube-api-access-7f4dx"} 0.000389135
kubelet_volume_metric_collection_duration_seconds{pod_name="tigera-operator-59d6fdcd79-hk6ml",pod_namespace="tigera-operator",volume_name="kube-api-access-bl2pr"} 0.000491883
kubelet_volume_metric_collection_duration_seconds{pod_name="tigera-operator-59d6fdcd79-hk6ml",pod_namespace="tigera-operator",volume_name="var-lib-calico"} 4.7391e-05
```

The current one using histogram:
```
# HELP kubelet_volume_stat_cal_duration_seconds [ALPHA] Duration in seconds to calculate volume stats
# TYPE kubelet_volume_stat_cal_duration_seconds histogram
kubelet_volume_stat_cal_duration_seconds_bucket{le="0.005"} 374
kubelet_volume_stat_cal_duration_seconds_bucket{le="0.01"} 374
kubelet_volume_stat_cal_duration_seconds_bucket{le="0.025"} 374
kubelet_volume_stat_cal_duration_seconds_bucket{le="0.05"} 374
kubelet_volume_stat_cal_duration_seconds_bucket{le="0.1"} 374
kubelet_volume_stat_cal_duration_seconds_bucket{le="0.25"} 374
kubelet_volume_stat_cal_duration_seconds_bucket{le="0.5"} 374
kubelet_volume_stat_cal_duration_seconds_bucket{le="1"} 374
kubelet_volume_stat_cal_duration_seconds_bucket{le="2.5"} 374
kubelet_volume_stat_cal_duration_seconds_bucket{le="5"} 380
kubelet_volume_stat_cal_duration_seconds_bucket{le="10"} 389
kubelet_volume_stat_cal_duration_seconds_bucket{le="+Inf"} 390
kubelet_volume_stat_cal_duration_seconds_sum 118.40980060400001
kubelet_volume_stat_cal_duration_seconds_count 390
```

Reply:
> Should this have any labels? i.e - getting stats can vary between volume backends and do we need to identify which storage provider we are dealing with?

Or adding a volume type? But I am not sure the easy way to group volume by fs type(emptydir\configMap\...)

Reply:
I was thinking more like volume type, but that information is not available in the stats calculator right now. The only way I can think of is to make that information available via `ListVolumesForPod` call from volume plugins. Also you will have to aggregate the numbers across different volumes of same volume type.

Using pod name, pod namespace etc is IMO too noisy and as you pointed out will cause an explosion in number of time series emitted by kubelet. But having a stat information without any kind of volume type information is also sort of not very useful.

Reply:
> I was thinking more like volume type, but that information is not available in the stats calculator right now. The only way I can think of is to make that information available via `ListVolumesForPod` call from volume plugins. Also you will have to aggregate the numbers across different volumes of same volume type.
> 
> Using pod name, pod namespace etc is IMO too noisy and as you pointed out will cause an explosion in number of time series emitted by kubelet. But having a stat information without any kind of volume type information is also sort of not very useful.

If we find out a way to use volume type, there is still a problem. 
Taking emptydirs as an example, most emptydirs are small and the time cost is tiny. What we want to expose to users is that some volume disk usage calculation is using too much time. (I will show you an example later. [TO BE Edited])


Another proposal would be `kubelet_slow_volume_metric_collection_duration_seconds` which means volume metric collection duration is more than 1s. For this metric, we only add a `guage` for slow ones. This will not be that noisy, but I am not sure if this is a general metric.

```
kubelet_slow_volume_metric_collection_duration_seconds{namespace="kube-system",pod="xxx-xxx",volume="big-data"} 1.2311
```

Reply:
If we only show slow metric collection, the metrics would be clean.
```
# TYPE kubelet_volume_slow_metric_collection_duration_seconds gauge
kubelet_volume_slow_metric_collection_duration_seconds{namespace="default",pod="pod-test",volume="data-test"} 6.800580388
kubelet_volume_slow_metric_collection_duration_seconds{namespace="default",pod="pod-test1",volume="data-test"} 2.241108193
```

Code like below:
```
                        defer func() {
+                               duration := servermetrics.SinceInSeconds(startTime)
                                // record the latest duration it took to get metrics for this volume
-                               servermetrics.VolumeStatCalDuration.WithLabelValues(s.pod.Namespace, s.pod.Name, name).Set(servermetrics.SinceInSeconds(startTime))
+                               // if the duration no more than 1 seconds, no need to show the metric
+                               if duration > 1 {
+                                       servermetrics.VolumeStatCalDuration.WithLabelValues(s.pod.Namespace, s.pod.Name, name).Set(duration)
+                               } else {
+                                       servermetrics.VolumeStatCalDuration.Delete(metrics.Labels{
+                                               "namespace": s.pod.Namespace,
+                                               "pod":       s.pod.Name,
+                                               "volume":    name,
+                                       })
+                               }
                        }()
```

Reply:
After some rethinking, I think logging it seems to be much simpler than the metrics.

Log a warning message if the duration is more than 1 second. Is this better?

Reply:
Sorry, I haven't been paying close attention to this.  This should probably be a histogram.  Gauge metrics are much harder to calculate percentiles over.  If you want it broken down by volume type, you can either record metrics from within each implementation of volume metrics (which allows you to only add metrics for implementations you are interested in), or add a Name function to the MetricsProvider interface.  If you are concerned about cardinality, consider reducing the number of buckets